charm/src/charm.ts

```diff
@@ -366,45 +380,235 @@ export class CharmManager {
+              if (sourceRefId) addMatchingCharm(sourceRefId);
+            } catch (err) {
+              console.debug("Error handling alias reference:", err);
             }
           }
-
-          // Look for direct cell references (cell + path properties)
+          
+          // Direct cell reference handling (for cases not caught by maybeGetCellLink)
           if (value.cell && value.path !== undefined) {
-            const cellId = getEntityId(value.cell);
-            if (cellId && cellId["/"]) {
-              // Found direct cell reference
-              addMatchingCharm(cellId);
+            try {
+              const cellId = getEntityId(value.cell);
+              if (cellId) addMatchingCharm(cellId);
+              
+              const sourceRefId = followSourceToResultRef(value.cell, new Set(), 0);
+              if (sourceRefId) addMatchingCharm(sourceRefId);
+            } catch (err) {
+              console.debug("Error handling direct cell reference:", err);
             }
           }
+          
+          // Safe recursive processing of arrays
+          if (Array.isArray(value)) {
+            for (let i = 0; i < value.length; i++) {
+              // Skip null/undefined items
+              if (value[i] == null) continue;
+              
+              // Skip items that might be cells to avoid Copy trap
+              if (typeof value[i] === 'object' && 
+                  (isCell(value[i]) || isDoc(value[i]) || isCellLink(value[i]))) {
+                try {
+                  // Process each cell directly
+                  processValue(value[i], parent, new Set([...visited]), depth + 1);
+                } catch (err) {
+                  console.debug(`Error processing special array item at index ${i}:`, err);
+                }
+                continue;
+              }
+              
+              // Process regular items
+              try {
+                processValue(value[i], parent, new Set([...visited]), depth + 1);
+              } catch (err) {
+                console.debug(`Error processing array item at index ${i}:`, err);
+              }
+            }
+          } else if (typeof value === 'object') {
+            // Process regular object properties
+            const keys = Object.keys(value);
+            for (let i = 0; i < keys.length; i++) {
+              const key = keys[i];
+              
+              // Skip properties that might be or contain Cell objects
+              if (key === 'sourceCell' || key === 'cell' || key === 'value' || 
+                  key === 'getAsCellLink' || key === 'getSourceCell') {
+                continue;
+              }
+              
+              try {
+                processValue(value[key], parent, new Set([...visited]), depth + 1);
+              } catch (err) {
+                console.debug(`Error processing object property '${key}':`, err);
+              }
+            }
+          }
```

Comment:
I don't really understand how or why, but I was triggering a the copy trap by naively traversing.